### ISSUE TITLE
Matrix grid doesn't always work with different central meridians

### ISSUE THREAD

**Issue by seisman** [GenericMappingTools/gmt#4335]
**Description of the problem**

The issue was first reported in https://github.com/GenericMappingTools/pygmt/issues/390, but was thought to be fixed by GMT PRs #3813 and #3829. The GMT master branch works for some projection parameters but doesn't always work.

**Full script that generated the error**

```python
import pygmt
fig = pygmt.Figure()

fig.grdimage("@earth_relief_01d_g", projection="Q123/0/15c", cmap="geo")

fig.shift_origin(yshift="8c")

dem = pygmt.datasets.load_earth_relief(resolution="01d", registration="gridline")
fig.grdimage(dem, projection="Q123/0/15c", cmap="geo")
fig.savefig("map.pdf")
```

**Output**

![image](https://user-images.githubusercontent.com/3974108/96204291-f00b0400-0f31-11eb-97fd-7cc1bf51b496.png)

**Image difference**
![image](https://user-images.githubusercontent.com/3974108/96204391-2fd1eb80-0f32-11eb-8f18-e64cbd0deed8.png)

In the PyGMT, we test several different projection parameters. Here are the ones that are known to fail:

- `Q123/0/15c`
- `Q123/30/15c`
- `Q180/0/15c`
- `Q180/30/15c`

Here are the ones that are known to work:

- `Q0/0/15c`
- `Q0/30/15c`

The same parameters work well for General Stereographic projection (i.e., changing **Q** to **S**).

Note: Those tests were originally written by @weiji14.

**Comment by PaulWessel** [comment]
Is this a gridline registration issue only or is it the same for pixel as well?

**Comment by seisman** [comment]
Same issue for pixel-registered grids (just tried `Q123/0/15c`).

**Comment by PaulWessel** [comment]
And just to be absolutely clear, the plot on the bottom is what you expect?  That rougher plot is what I get with long=0 which you said works, so just checking.

**Comment by seisman** [comment]
> And just to be absolutely clear, the plot on the bottom is what you expect?

Yes, the bottom one uses the earth relief grid file, which gives the same plot as GMT CLI.

**Comment by PaulWessel** [comment]
So what happens is this: We have a read-only grid and it has a certain central meridian.  This kicks in:

```
if (got_z_grid && (gmt_whole_earth (GMT, Grid_orig->header->wesn, wesn) == 1))
	need_to_project = true;	/* This can only happen if reading a remote global geographic grid and the central meridian differs from that of the projection */
```

So now we go the way of resampling the images - hence the smoother apparance.  The other image is not resampled, just plotted, but handles the change in longitude.  I will see if I can make it work the same.  You would think that should be doable, but the big difference is that the file is read in and the matrix is rotated to fit the -R while in the xarray case the pixels are still -180/180.  So not the same case.

**Comment by joa-quim** [comment]
What is the status of this? Mane related changes have been committed meawhile.

**Comment by seisman** [comment]
I'll find some time to check it.

**Comment by joa-quim** [comment]
Another ping

**Comment by seisman** [comment]
Still fail with the master branch, for `Q123/0/15c` and `Q180/0/15c`

**Comment by joa-quim** [comment]
A CLI example?

**Comment by seisman** [comment]
The bug only exists when passing virtual files from wrappers.
